state/sequence.go

```diff
@@ -38,3 +43,128 @@ func (s *State) sequence(name string) (int, error) {
+	defer closer()
+	updater := &dbSeqUpdater{
+		coll: sequences,
+		id:   st.docID(name),
+		name: name,
+	}
+	return updateSeqWithMin(updater, minVal)
+}
+
+// seqUpdater abstracts away the database operations required for
+// updating a sequence.
+type seqUpdater interface {
+	read() (int, error)
+	create(value int) (bool, error)
+	set(expected, next int) (bool, error)
+}
+
+const maxSeqRetries = 20
+
+// updateSeqWithMin implements the abstract logic for incrementing a
+// database backed sequence in a concurrency aware way.
+//
+// It is complicated because MongoDB's atomic update primitives don't
+// provide a way to upsert a counter while also providing an initial
+// value.  Instead, a number of database operations are used for each
+// sequence update, relying on the atomicity guarantees that MongoDB
+// offers. Optimistic database updates are attempted with retries when
+// contention is observed.
+func updateSeqWithMin(sequence seqUpdater, minVal int) (int, error) {
+	for try := 0; try < maxSeqRetries; try++ {
+		curVal, err := sequence.read()
+		if err != nil {
+			return -1, errors.Annotate(err, "could not read sequence")
+		}
+		if curVal == 0 {
+			// No sequence document exists, create one.
+			ok, err := sequence.create(minVal + 1)
+			if err != nil {
+				return -1, errors.Annotate(err, "could not create sequence")
+			}
+			if ok {
+				return minVal, nil
+			}
+			// Someone else created the sequence document at the same
+			// time, try again.
+		} else {
+			// Increment an existing sequence document, respecting the
+			// minimum value provided.
+			nextVal := curVal + 1
+			if nextVal < minVal {
+				nextVal = minVal + 1
+			}
+			ok, err := sequence.set(curVal, nextVal)
+			if err != nil {
+				return -1, errors.Annotate(err, "could not set sequence")
+			}
+			if ok {
+				return nextVal - 1, nil
+			}
+			// Someone else incremented the sequence at the same time,
+			// try again.
+		}
+	}
+	return -1, errors.New("too much contention while updating sequence")
+}
+
+// dbSeqUpdater implements seqUpdater.
+type dbSeqUpdater struct {
+	coll mongo.Collection
+	id   string
+	name string
+}
+
+func (su *dbSeqUpdater) read() (int, error) {
+	var doc bson.M
+	err := su.coll.FindId(su.id).One(&doc)
+	if errors.Cause(err) == mgo.ErrNotFound {
+		return 0, nil
+	} else if err != nil {
+		return -1, errors.Trace(err)
+	}
+	return doc["counter"].(int), nil
+}
+
+func (su *dbSeqUpdater) create(value int) (bool, error) {
+	err := su.coll.Writeable().Insert(bson.M{
+		"_id":     su.id,
+		"name":    su.name,
+		"counter": value,
+	})
+	if mgo.IsDup(errors.Cause(err)) {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Trace(err)
+	}
+	return true, nil
+}
+
+func (su *dbSeqUpdater) set(expected, next int) (bool, error) {
+	err := su.coll.Writeable().Update(
+		bson.M{"_id": su.id, "counter": expected},
+		bson.M{"$set": bson.M{"counter": next}},
+	)
+	if errors.Cause(err) == mgo.ErrNotFound {
+		return false, nil
```

Comment:
The update query includes the expected current value of the sequence. The update will fail if ErrNotFound if someone update the sequence, allowing detection of concurrent updates.